qwerty_ast/src/typecheck.rs

```diff
@@ -0,0 +1,456 @@
+                            },
+                            span: None,
+                        });
+                    }
+                    Ok((**out_ty).clone())
+                }
+                _ => Err(TypeError {
+                    kind: TypeErrorKind::NotCallable(format!("{:?}", rhs_ty)),
+                    span: None,
+                }),
+            }
+        }
+
+        Expr::Measure { basis, span: _ } => {
+            // Qwerty: measurement returns classical result; basis must be valid.
+            typecheck_basis(basis, env)?; //  is it a legal quantum basis?
+
+            Ok(Type::RegType {
+                elem_ty: RegKind::Bit,
+                dim: 1, // TODO: Make dynamic based on basis (Check with Austin about its validation)
+                // Self Note (verify with Austin): Currently this measurement returns a single classical bit
+                // But in real quantum programs, we might measure multiple qubits at once (e.g measuring a register of 3 qubits gives you 3 classical bits, ryt?)
+                // The number of bits returned should depend on the size/dimension of the basis being measured
+                // So, make 'dim' reflect the actual number of qubits measured, as determined by the basis argument
+                // Need to understand and work on dimensions.. Discuss bro!
+            })
+        }
+
+        Expr::Discard { span: _ } => Ok(Type::UnitType),
+
+        // A tensor (often a tensor product) means combining multiple quantum states or registers into a larger, composite system
+        Expr::Tensor { vals, span: _ } => {
+            // => All sub-expressions in a tensor must have the same type (e.g. all are qubits, or all are bits)
+            // Tensor([Qubit, Qubit, Qubit])
+            let mut t = None;
+            for v in vals {
+                let vty = typecheck_expr(v, env)?;
+
+                if let Some(prev) = &t {
+                    if &vty != prev {
+                        // if current_type doesn't match with prev_type, error!
+                        return Err(TypeError {
+                            kind: TypeErrorKind::MismatchedTypes {
+                                expected: format!("{:?}", prev),
+                                found: format!("{:?}", vty),
+                            },
+                            span: None,
+                        });
+                    }
+                }
+                t = Some(vty);
+            }
+            Ok(t.unwrap_or(Type::UnitType))
+        }
+
+        Expr::BasisTranslation { bin, bout, span: _ } => {
+            // TODO: Ensure translation is between compatible bases.
+            /*
+            0) ASK Austin!
+            1) Typecheck both bases (already done)
+            2) Extract relevant info from each basis (e.g. dimension, type).
+            3) Compare the properties we care about (e.g. same dimension, same qubit type).
```

Comment:
I need to come up with an algorithm for span checking. There is an algorithm in this paper https://dl.acm.org/doi/10.1145/3696443.3708966 but it is already outdated now